File: ios/MRT/DesignSystem/Components/GHDiffView.swift
import SwiftUI

struct GHDiffLine: Identifiable {
    enum Kind {
        case addition
        case deletion
        case context
    }

    let id = UUID()
    let kind: Kind
    let content: String
}

struct GHDiffView: View {
    let title: String
    let lines: [GHDiffLine]

    var body: some View {
        GHCard {
            Text(title)
                .font(GHTypography.bodySm)
                .foregroundStyle(GHColors.textSecondary)

            VStack(spacing: 0) {
                ForEach(lines) { line in
                    HStack(alignment: .top, spacing: GHSpacing.md) {
                        Rectangle()
                            .fill(gutterColor(for: line.kind))
                            .frame(width: 3)

                        Text(line.content)
                            .font(GHTypography.code)
                            .foregroundStyle(GHColors.textPrimary)
                            .frame(maxWidth: .infinity, alignment: .leading)
                    }
                    .padding(.vertical, GHSpacing.xs)
                    .padding(.horizontal, GHSpacing.sm)
                    .background(backgroundColor(for: line.kind))
                }
            }
            .background(GHColors.bgPrimary)
            .clipShape(RoundedRectangle(cornerRadius: GHRadius.md))
        }
    }

    private func gutterColor(for kind: GHDiffLine.Kind) -> Color {
        switch kind {
        case .addition:
            return GHColors.accentGreen
        case .deletion:
            return GHColors.accentRed
        case .context:
            return GHColors.borderDefault
        }
    }

    private func backgroundColor(for kind: GHDiffLine.Kind) -> Color {
        switch kind {
        case .addition:
            return GHColors.accentGreen.opacity(0.08)
        case .deletion:
            return GHColors.accentRed.opacity(0.08)
        case .context:
            return GHColors.bgPrimary
        }
    }
}

#Preview {
    ZStack {
        GHColors.bgPrimary.ignoresSafeArea()
        GHDiffView(
            title: "Placeholder Diff",
            lines: [
                .init(kind: .context, content: "@@ -1,3 +1,3 @@"),
                .init(kind: .deletion, content: "- old line"),
                .init(kind: .addition, content: "+ new line"),
            ]
        )
        .padding()
    }
}

File: android/app/src/main/java/com/mrt/app/designsystem/components/GHDiffView.kt
package com.mrt.app.designsystem.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.mrt.app.designsystem.theme.GHColors
import com.mrt.app.designsystem.theme.GHRadii
import com.mrt.app.designsystem.theme.GHSpacing
import com.mrt.app.designsystem.theme.GHType

enum class GHDiffKind {
    Added,
    Removed,
    Context,
}

data class GHDiffLine(
    val kind: GHDiffKind,
    val content: String,
)

@Composable
fun GHDiffView(
    title: String,
    lines: List<GHDiffLine>,
    modifier: Modifier = Modifier,
) {
    Surface(
        modifier = modifier.fillMaxWidth(),
        color = GHColors.BgSecondary,
        shape = androidx.compose.foundation.shape.RoundedCornerShape(GHRadii.Lg),
        border = BorderStroke(1.dp, GHColors.BorderDefault),
    ) {
        Column(
            modifier = Modifier.padding(GHSpacing.Lg),
            verticalArrangement = Arrangement.spacedBy(GHSpacing.Sm),
        ) {
            Text(
                text = title,
                style = GHType.BodySm,
                color = GHColors.TextPrimary,
                fontWeight = FontWeight.SemiBold,
            )
            lines.forEach { line ->
                val prefix = when (line.kind) {
                    GHDiffKind.Added -> "+"
                    GHDiffKind.Removed -> "-"
                    GHDiffKind.Context -> " "
                }
                val color = when (line.kind) {
                    GHDiffKind.Added -> GHColors.AccentGreen
                    GHDiffKind.Removed -> GHColors.AccentRed
                    GHDiffKind.Context -> GHColors.TextSecondary
                }
                val background = when (line.kind) {
                    GHDiffKind.Added -> GHColors.AccentGreen.copy(alpha = 0.08f)
                    GHDiffKind.Removed -> GHColors.AccentRed.copy(alpha = 0.08f)
                    GHDiffKind.Context -> Color.Transparent
                }
                Text(
                    text = "$prefix ${line.content}",
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = GHSpacing.Sm)
                        .background(background, androidx.compose.foundation.shape.RoundedCornerShape(GHRadii.Sm))
                        .padding(horizontal = GHSpacing.Sm, vertical = GHSpacing.Xs),
                    style = GHType.Code,
                    color = color,
                )
            }
        }
    }
}
